pkg/kube/namespaces.go

```diff
@@ -0,0 +1,218 @@
+package kube
+
+import (
+	"context"
+	"net"
+	"os"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/rancher/prometheus-auth/pkg/data"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/oauth2"
+	"golang.org/x/sync/singleflight"
+	authorization "k8s.io/api/authorization/v1"
+	core "k8s.io/api/core/v1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	authorizationTyped "k8s.io/client-go/kubernetes/typed/authorization/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+)
+
+type Namespaces interface {
+	Query(token string) data.Set
+}
+
+type namespaces struct {
+	singleflight.Group
+
+	ctx              context.Context
+	secretIndexer    cache.Indexer
+	namespaceIndexer cache.Indexer
+}
+
+func (n *namespaces) Query(token string) data.Set {
+	ret, err, _ := n.Do(token, func() (interface{}, error) {
+		ret := data.Set{}
+
+		tokenNamespace, err := n.validate(token)
+		if err != nil {
+			return ret, err
+		}
+
+		nsObj, exist, _ := n.namespaceIndexer.GetByKey(tokenNamespace)
+		if !exist {
+			return ret, errors.New("unknown namespace of token")
+		}
+
+		ns := toNamespace(nsObj)
+		if ns.DeletionTimestamp != nil {
+			return ret, errors.New("deleting namespace of token")
+		}
+
+		projectID, exist := getProjectID(ns)
+		if !exist {
+			return ret, errors.New("unknown project of token")
+		}
+
+		nsList, err := n.namespaceIndexer.ByIndex("namespaceByProjectID", projectID)
+		if err != nil {
+			return ret, errors.Annotatef(err, "invalid project")
+		}
+
+		for _, nsObj := range nsList {
+			ns := toNamespace(nsObj)
+			ret[ns.Name] = struct{}{}
+		}
+		return ret, nil
+	})
+	if err != nil {
+		log.Warnln("failed to query Namespaces", errors.ErrorStack(err))
+	}
+
+	return ret.(data.Set)
+}
+
+func (n *namespaces) validate(token string) (string, error) {
+	secList, err := n.secretIndexer.ByIndex("secretByToken", token)
+	if err != nil || len(secList) != 1 {
+		return "", errors.Annotatef(err, "unknown token")
+	}
+
+	sec := toSecret(secList[0])
+	if sec.DeletionTimestamp != nil {
+		return "", errors.New("deleting token")
+	}
+
+	authorizationClient, err := authorizationTyped.NewForConfig(clusterConfigByToken(token))
+	if err != nil {
+		return "", errors.Annotatef(err, "invalid token")
+	}
+
+	sar := &authorization.SelfSubjectAccessReview{
```

Comment:
Can you use SubjectAccessReview, not SelfSubjectAccessReview.  The idea being that you construct one client with the prometheus-auth service account credentials and then use that same client to check other subjects.  The problem with this approach is that creating new clients on demand is dangerous due to http connection pooling.  You might end up leaking connections.

Reply:
darren gave me this format for supplying a service account in the SubjectAccessReview user fileld. i am hoping to have time to try it out myself, but I may not, so you can @thxCode:
 `"system:serviceaccount:kube-system:default"`

Reply:
@cjellick , Got.